eventuate-messaging-kafka-common/src/main/java/io/eventuate/messaging/kafka/common/EventuateKafkaMultiMessageConverter.java

```diff
@@ -151,11 +224,24 @@ private int estimatedStringSizeInBytes(String s) {
 
       MultiMessageEncoder multiMessageEncoder = new MultiMessageEncoder().wrapAndApplyHeader(buffer, 0, messageHeaderEncoder);
 
+      MultiMessageEncoder.HeadersEncoder headersEncoder = multiMessageEncoder.headersCount(headers.size());
+      headers.forEach(header -> headersEncoder.next().key(header.getKey()).value(header.getValue()));
+
       MultiMessageEncoder.MessagesEncoder messagesEncoder = multiMessageEncoder.messagesCount(messagesToWrite.size());
+      messagesToWrite.forEach(message -> {
+        messagesEncoder.next();
+
+        MultiMessageEncoder.MessagesEncoder.HeadersEncoder messageHeadersEncoder =
+                messagesEncoder.headersCount(message.getHeaders().size());
+
+        message.getHeaders().forEach(header -> messageHeadersEncoder.next().key(header.getKey()).value(header.getValue()));
```

Comment:
Would a for loop be more efficient, e.g. not cause heap allocation of lambda? if so, use it

Reply:
it should be